src/utils/colorManipulation.test.ts

```diff
@@ -0,0 +1,231 @@
+// `getAverageColor` function are in the same describe.
+describe("colorManipulation", () => {
+  // I declare these vary early on so we can refer to them in all future if/test
+  // blocks. This `beforeEach` hook will run before every test. This means that
+  // for every test that runs, we get a fresh canvas and context so there isn't
+  // any test influencing others.
+  let canvas: HTMLCanvasElement;
+  let ctx: CanvasRenderingContext2D;
+
+  beforeEach(() => {
+    canvas = document.createElement("canvas");
+    canvas.width = 100;
+    canvas.height = 100;
+    ctx = canvas.getContext("2d", { willReadFrequently: true })!;
+    ctx.clearRect(0, 0, canvas.width, canvas.height);
+  });
+
+  describe("getAverageColor", () => {
+    // Use a `test` or `it` block to run an actual test. The string is just a brief
+    // description of the scenario or specific code path you're testing. The
+    // function is the actual test. The `it` and your string should read like a sentence.
+    it("averages one color as itself", () => {
+      // Setting up the dummy data that my test can use
+      ctx.fillStyle = "rgb(255, 0, 0)";
+      ctx.fillRect(0, 0, canvas.width, canvas.height);
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+
+      // Determine what my test should return by hand
+      const expected = [255, 0, 0, 255];
+
+      // Run my code and get it's actual output
+      const actual = getAverageColor(frame, 0, 0, canvas.height, canvas.width);
+
+      // Use an expect to actual run assertions with the matchers that jest provides
+      expect(actual).toEqual(expected);
+    });
+
+    it("averages two colors", () => {
+      ctx.fillStyle = "rgb(0, 0, 0)";
+      ctx.fillRect(0, 0, canvas.width / 2, canvas.height);
+      ctx.fillStyle = "rgb(255, 255, 255)";
+      ctx.fillRect(canvas.width / 2, 0, canvas.width, canvas.height);
+
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      const expected = [128, 128, 128, 255];
+      const actual = getAverageColor(frame, 0, 0, canvas.height, canvas.width);
+      expect(actual).toEqual(expected);
+    });
+
+    it("averages four colors", () => {
+      // top left quadrant
+      ctx.fillStyle = "rgb(255, 0, 0)";
+      ctx.fillRect(0, 0, canvas.width / 4, canvas.height / 4);
+
+      // top right quadrant
+      ctx.fillStyle = "rgb(0, 255, 0)";
+      ctx.fillRect(canvas.width / 2, 0, canvas.width / 4, canvas.height / 4);
+
+      // bottom left quadrant
+      ctx.fillStyle = "rgb(0, 0, 255)";
+      ctx.fillRect(0, canvas.height / 2, canvas.width / 4, canvas.height / 4);
+
+      // bottom right quadrant
+      ctx.fillStyle = "rgb(0, 0, 0)";
+      ctx.fillRect(
+        canvas.width / 2,
+        canvas.height / 2,
+        canvas.width / 4,
+        canvas.height / 4,
+      );
+
+      const frame = ctx.getImageData(0, 0, canvas.width, canvas.height);
+      const expected = [16, 16, 16, 255];
```

Comment:
good call, gonna look over this test again

Reply:
looked over it. the canvas is 100x100 and there's four quadrants so 4 25x25 squares. each square has 625 pixels so  the sum total color of each quadrant is 255 * # of pixels giving 159,375. this means the sum RGB of the entire canvas is `[156375, 156375, 156375, 255]`. we then divide this by the number of pixels which is 10k in a 100x100. this brings the average RGBA of the canvas to `[15.9375, 15.9375, 15.9375, 255]` which will round to 16. 

worth noting that we end up with this because we compute the average color by using the number of pixels versus the number of unique colors and doing some weight preprocessing like that. they should lead to roughly the same answer so I'm not too worried

thanks for the opportunity to explain

Reply:
this makes sense now, i did not account the number of pixels playing a factor